### PR TITLE
Fix auditlog format

### PR DESCRIPTION
**Pull request by moelsayed**
https://github.com/kyma-incubator/reconciler/issues/291

### CODE REVIEW

File: cmd/mothership/mothership/start/auditlog.go [hunk 1]
```diff
@@ -107,8 +131,50 @@ func auditLogRequest(w http.ResponseWriter, r *http.Request, l *zap.Logger) {
 			return
 		}
 		r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
-		l = l.With(zap.String("requestBody", string(reqBody)))
+		logData.RequestBody = string(reqBody)
+	}
+
+	if ip, ok := os.LookupEnv("POD_IP"); ok {
+		logData.IP = ip
```
Comment (a-thaler) [inline cmd/mothership/mothership/start/auditlog.go:138]: I guess the IP should be the source IP of the incoming request as this is relevant info for the auditing

File: cmd/mothership/mothership/start/auditlog.go [hunk 2]
```diff
@@ -107,8 +131,50 @@ func auditLogRequest(w http.ResponseWriter, r *http.Request, l *zap.Logger) {
...
+		With(zap.String("category", "audit.security-events")). // comply with required log backend format
+		Info("")
+}
+
+func getJWTPayload(r *http.Request) (string, error) {
+	jwtHeader := r.Header.Get(XJWTHeaderName)
+	if len(jwtHeader) == 0 {
+		return "", nil
+	}
+	decodedSeg, err := base64.RawURLEncoding.DecodeString(jwtHeader)
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog.go:165]: I got 500 internal server error while passing JWT token.
You should use jwt parser or split the token by `.` and then decode it.

**Review by dbadura — CHANGES_REQUESTED** [review]
Please add unit test for `auditlog`

File: cmd/mothership/mothership/start/auditlog.go [hunk 3]
```diff
@@ -59,17 +63,28 @@ func NewLoggerWithFile(logFile string) (*zap.Logger, error) {
 	), err
 }
 
-func NewAuditLoggerMiddelware(l *zap.Logger) func(http.Handler) http.Handler {
+func NewAuditLoggerMiddelware(l *zap.Logger, o *Options) func(http.Handler) http.Handler {
```
Comment (tobiscr) [inline cmd/mothership/mothership/start/auditlog.go:66]: Would recommend to make it a private function as it shouldn't be used outside of this pkg.

**Review by tobiscr — CHANGES_REQUESTED** [review]

File: cmd/mothership/mothership/start/auditlog.go [hunk 4]
```diff
@@ -79,22 +93,32 @@ func auditLogRequest(w http.ResponseWriter, r *http.Request, l *zap.Logger) {
...
-			server.SendHTTPError(w, http.StatusInternalServerError, &keb.HTTPErrorResponse{
-				Error: errors.Wrap(err, fmt.Sprintf("Failed to parse %s header content ", XJWTHeaderName)).Error(),
-			})
-			return
-		}
+	logData := data{
+		ContractVersion: contractV,
+		Method:          r.Method,
+		URI:             r.RequestURI,
+		User:            "UNKOWEN_USER",
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog.go:100]: typo

File: cmd/mothership/mothership/start/auditlog_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+// validateLog ensures that all required fields in the log message are set and valid. If any of these is missing the audit log backend will not accept/process our logs
+func validateLog(logMsg, method string, useJWT bool) error {
+	l := &log{}
+	if err := json.Unmarshal([]byte(logMsg), l); err != nil {
+		return err
+	}
+	if l.Time == "" ||
+		l.UUID == "" ||
+		l.User == "" ||
+		l.Data == "" ||
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog_test.go:156]: Validate if filled data is correct.

File: cmd/mothership/mothership/start/auditlog_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		EncoderConfig: zapcore.EncoderConfig{
+			MessageKey:   "",
+			LevelKey:     "level",
+			EncodeLevel:  zapcore.CapitalLevelEncoder,
+			TimeKey:      "time",
+			EncodeTime:   zapcore.ISO8601TimeEncoder,
+			EncodeCaller: zapcore.ShortCallerEncoder,
+		},
+	}
+	logger, _ := cfg.Build()
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog_test.go:54]: Don't swallow the error. Use `require.NoError(...)`

File: cmd/mothership/mothership/start/auditlog_test.go (hunk 5 above)
Comment (moelsayed) [reply cmd/mothership/mothership/start/auditlog_test.go:157]: The non-generated fields are checked later in the function, like tenant, user, data..etc.

File: cmd/mothership/mothership/start/auditlog_test.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+		l.Category == "", "empty log field: %#v", l)
+
+	require.Equalf(t, tenantID, l.Tenant, "invalid log tenantID: expected: %s, got: %s", tenantID, l.Tenant)
+
+	if useJWT {
+		require.Equalf(t, jwtPayloadSub, l.User, "invalid user: expected: %s, got: %s", jwtPayloadSub, l.User)
+	}
+	if method == http.MethodPost {
+		d := &data{}
+		if err := json.Unmarshal([]byte(l.Data), d); err != nil {
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog_test.go:165]: require.NoError is enought.

File: cmd/mothership/mothership/start/auditlog_test.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+		if testCase.expectFail {
+			require.Equalf(t, http.StatusInternalServerError, w.Result().StatusCode,
+				"expected http status: %v, got: %v",
+				http.StatusInternalServerError, w.Result().StatusCode)
+		} else {
+			t.Log(output.String())
+			validateLog(t, output.String(), testCase.method, testCase.jwtHeader != "")
+		}
+		// clean the log sink
+		output.Reset()
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog_test.go:140]: `defer output.Reset()`, before `WHEN`?

File: cmd/mothership/mothership/start/auditlog_test.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	// build test logger
+	logger, output := testLoggerWithOutput(t)
+	// build reconciler options
+	o := NewOptions(&cli.Options{})
+	o.AuditLogTenantID = tenantID
+
+	for _, testCase := range testCases {
+		// GIVEN
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog_test.go:111]: t.Run(.....)

**Review by dbadura — COMMENTED** [review]
Minor things

File: cmd/mothership/mothership/start/auditlog_test.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+	require.Equalf(t, clientIP, l.IP, "invalid log IP: expected: %s, got: %s", clientIP, l.IP)
+
+	if useJWT {
+		require.Equalf(t, jwtPayloadSub, l.User, "invalid user: expected: %s, got: %s", jwtPayloadSub, l.User)
+	}
+	if method == http.MethodPost {
+		d := &data{}
+		if err := json.Unmarshal([]byte(l.Data), d); err != nil {
+			require.NoError(t, err)
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog_test.go:175]:
```suggestion
		err := json.Unmarshal([]byte(l.Data), d)
		require.NoError(t, err)
```

File: cmd/mothership/mothership/start/auditlog.go [hunk 11]
```diff
@@ -107,8 +132,53 @@ func auditLogRequest(w http.ResponseWriter, r *http.Request, l *zap.Logger) {
...
 		r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
-		l = l.With(zap.String("requestBody", string(reqBody)))
+		logData.RequestBody = string(reqBody)
+	}
+
+	if ip := r.Header.Get(ExternalAddressHeaderName); ip != "" {
+		logData.IP = ip
+	} else {
+		o.Logger().Debug(fmt.Sprintf("empty %s header", ExternalAddressHeaderName))
+	}
```
Comment (dbadura) [inline cmd/mothership/mothership/start/auditlog.go:142]:
```suggestion
	ip := r.Header.Get(ExternalAddressHeaderName)
	if ip == "" {
		o.Logger().Debug(fmt.Sprintf("empty %s header", ExternalAddressHeaderName))
	}
	logData.IP = ip
```

**Review by dbadura — DISMISSED** [review]

**Review by dbadura — APPROVED** [review]

**Review by tobiscr — APPROVED** [review]

**Pull request merged**